[MRG] Extract some cross_val_predict/score functionality

caioaao opened the pull request:
This should provide an unobtrusive path for solving #8043 
Trying to make a more step-by-step fix here so it doesn't mess with 0.19 release schedule. This PR would allow for `_split_score` to be used in other places (or directly if someone is impatient).

The end goal here is to be able to reuse this chunk of code to create `backtest_predict` and `backtest_score` for time series splits.

caioaao committed:
Extract some cross_val_* functionality

Solves #8043

caioaao commented:
k. I'll wait for it to be merged. It has a clear gain: today scikit learn
has very limited functionality to deal with time series data. extracting
stuff like this would make it easier to write an backtesting validation API
that's equivalent to the current cross validation.

Em ter, 13 de jun de 2017 às 19:37, Joel Nothman <notifications@github.com>
escreveu:

> as we would like to get #7388 into the next release, it would probably be
> unwise to consider something attacking the same code without clear gain
>
> On 14 Jun 2017 2:39 am, "Caio Oliveira" <notifications@github.com> wrote:
>
> > This should make an unobtrusive wat for solving #8043
> > <https://github.com/scikit-learn/scikit-learn/issues/8043>
> > Trying to make a more step-by-step fix here so it doesn't mess with 0.19
> > release schedule. This PR would allow for _split_score to be used in
> > other places (or directly if someone is impatient).
> >
> > The end goal here is to be able to reuse this chunk of code to create
> > backtest_predict and backtest_score for time series splits.
> > ------------------------------
> > You can view, comment on, or merge this pull request online at:
> >
> > https://github.com/scikit-learn/scikit-learn/pull/9117
> > Commit Summary
> >
> > - Extract some cross_val_* functionality
> >
> > File Changes
> >
> > - *M* sklearn/model_selection/_validation.py
> > <https://github.com/scikit-learn/scikit-learn/pull/9117/files#diff-0>
> > (233)
> >
> > Patch Links:
> >
> > - https://github.com/scikit-learn/scikit-learn/pull/9117.patch
> > - https://github.com/scikit-learn/scikit-learn/pull/9117.diff
> >
> > —
> > You are receiving this because you are subscribed to this thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/scikit-learn/scikit-learn/pull/9117>, or mute the
> > thread
> > <
> https://github.com/notifications/unsubscribe-auth/AAEz646DQmFyY_EtTr0gyvSrWl_t_Fauks5sDrtTgaJpZM4N4wbm
> >
> > .
> >
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9117#issuecomment-308268467>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABVNFnchNeOHTkyTs_gTIvhLn_-HqSpBks5sDw8fgaJpZM4N4wbm>
> .
>

jnothman commented:
Please provide usage examples

On 14 Jun 2017 8:45 am, "Caio Oliveira" <notifications@github.com> wrote:

> k. I'll wait for it to be merged. It has a clear gain: today scikit learn
> has very limited functionality to deal with time series data. extracting
> stuff like this would make it easier to write an backtesting validation API
> that's equivalent to the current cross validation.
>
> Em ter, 13 de jun de 2017 às 19:37, Joel Nothman <notifications@github.com
> >
> escreveu:
>
> > as we would like to get #7388 into the next release, it would probably be
> > unwise to consider something attacking the same code without clear gain
> >
> > On 14 Jun 2017 2:39 am, "Caio Oliveira" <notifications@github.com>
> wrote:
> >
> > > This should make an unobtrusive wat for solving #8043
> > > <https://github.com/scikit-learn/scikit-learn/issues/8043>
> > > Trying to make a more step-by-step fix here so it doesn't mess with
> 0.19
> > > release schedule. This PR would allow for _split_score to be used in
> > > other places (or directly if someone is impatient).
> > >
> > > The end goal here is to be able to reuse this chunk of code to create
> > > backtest_predict and backtest_score for time series splits.
> > > ------------------------------
> > > You can view, comment on, or merge this pull request online at:
> > >
> > > https://github.com/scikit-learn/scikit-learn/pull/9117
> > > Commit Summary
> > >
> > > - Extract some cross_val_* functionality
> > >
> > > File Changes
> > >
> > > - *M* sklearn/model_selection/_validation.py
> > > <https://github.com/scikit-learn/scikit-learn/pull/9117/files#diff-0>
> > > (233)
> > >
> > > Patch Links:
> > >
> > > - https://github.com/scikit-learn/scikit-learn/pull/9117.patch
> > > - https://github.com/scikit-learn/scikit-learn/pull/9117.diff
> > >
> > > —
> > > You are receiving this because you are subscribed to this thread.
> > > Reply to this email directly, view it on GitHub
> > > <https://github.com/scikit-learn/scikit-learn/pull/9117>, or mute the
> > > thread
> > > <
> > https://github.com/notifications/unsubscribe-auth/AAEz646DQmFyY_
> EtTr0gyvSrWl_t_Fauks5sDrtTgaJpZM4N4wbm
> > >
> > > .
> > >
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/scikit-learn/scikit-learn/pull/9117#
> issuecomment-308268467>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/ABVNFnchNeOHTkyTs_
> gTIvhLn_-HqSpBks5sDw8fgaJpZM4N4wbm>
> > .
> >
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9117#issuecomment-308270192>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6yVUuoynxnWupQlDPoAekJg3ubRwks5sDxEmgaJpZM4N4wbm>
> .
>

caioaao commented:
I'm sorry, but I believe I just did :thinking: but here's a more concrete example:

First, we can define `backtest_score`

```python
from sklearn.model_selection._validation import _split_score
from sklearn.model_selection import TimeSeriesSplit
from sklearn.utils import indexable


def time_series_split_check(split):
    # would be the equivalent of `cv_check`. would check if indexes are
    # increasing, etc...
    return split


def backtest_score(estimator, X, y=None, groups=None, scoring=None, split=None,
                   n_jobs=1, verbose=0, fit_params=None,
                   pre_dispatch='2*n_jobs'):
    """Same functionallity of `cross_val_score`, but with time series data"""
    X, y, groups = indexable(X, y, groups)
    split = time_series_split_check(split)
    return _split_score(estimator, split, X, y=y, groups=groups,
                        scoring=scoring, n_jobs=1, verbose=0,
                        fit_params=fit_params, pre_dispatch=pre_dispatch)
```

and then use it

```python
from sklearn.linear_model import LogisticRegression
from sklearn.datasets import make_classification

X, y = make_classification(n_samples=2000, n_features=20, n_informative=2,
                           n_redundant=2, n_repeated=0, n_classes=2,
                           n_clusters_per_class=2, weights=None, flip_y=0.01,
                           class_sep=1.0, hypercube=True, shift=0.0, scale=1.0,
                           shuffle=True, random_state=None)

split = TimeSeriesSplit(n_splits=3)
scores = backtest_score(LogisticRegression(), X, y, split=split,
                        scoring='neg_log_loss')
print(scores)
```

now imagine sharing the rest of the `cross_val_score` functionality with this... 

PS: `backtest_predict` would be trickier, so I don't have a clear idea about how one would implement it

jnothman commented:
so the difference is substituting for check_cv? Why not just create some kind of `class TimeSeriesAutoCV` which analyses the data and delegates when `.split` is called?

caioaao commented:
I don't know if I got that. do you mean `TimeSeriesAutoCV` would return a split compatible with `check_cv`?

I thought of creating another set of functions to work with time series data because it doesn't seem to play well with cross validation. As one evidence we can see the discussion in #8043, but, generally speaking, I thing a time series split is different enough to deserve its own set of specialized API. We could have other types of splits that take the temporal dependecy into account and they probably won't be compatible with CV functions.

jnothman commented:
I'm not altogether convinced this is the right long-term fix.

On 14 June 2017 at 12:41, Caio Oliveira <notifications@github.com> wrote:

> I don't know if I got that. do you mean TimeSeriesAutoCV would return a
> split compatible with check_cv?
>
> I thought of creating another set of functions to work with time series
> data because it doesn't seem to play well with cross validation. As one
> evidence we can see the discussion in #8043
> <https://github.com/scikit-learn/scikit-learn/issues/8043>, but,
> generally speaking, I thing a time series split is different enough to
> deserve its own set of specialized API. We could have other types of splits
> that take the temporal dependecy into account and they probably won't be
> compatible with CV functions.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9117#issuecomment-308303536>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6-yx7-0Q9Q6uXd_MGRbt9b5lqmA6ks5sD0hlgaJpZM4N4wbm>
> .
>

caioaao commented:
Me neither, but seems the most aligned with the current scikit-learn API.

Em ter, 13 de jun de 2017 às 23:46, Joel Nothman <notifications@github.com>
escreveu:

> I'm not altogether convinced this is the right long-term fix.
>
> On 14 June 2017 at 12:41, Caio Oliveira <notifications@github.com> wrote:
>
> > I don't know if I got that. do you mean TimeSeriesAutoCV would return a
> > split compatible with check_cv?
> >
> > I thought of creating another set of functions to work with time series
> > data because it doesn't seem to play well with cross validation. As one
> > evidence we can see the discussion in #8043
> > <https://github.com/scikit-learn/scikit-learn/issues/8043>, but,
> > generally speaking, I thing a time series split is different enough to
> > deserve its own set of specialized API. We could have other types of
> splits
> > that take the temporal dependecy into account and they probably won't be
> > compatible with CV functions.
> >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/scikit-learn/scikit-learn/pull/9117#issuecomment-308303536
> >,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AAEz6-yx7-0Q9Q6uXd_MGRbt9b5lqmA6ks5sD0hlgaJpZM4N4wbm
> >
> > .
> >
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9117#issuecomment-308304085>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABVNFkkjpm1RXfBU5bn27zkmjNWNUmd9ks5sD0mJgaJpZM4N4wbm>
> .
>
